Loader support and initial documentation

goldshtn opened the pull request:
This patchset introduces a Java loader tool that helps inject the agent into
a running Java process, and send the start, stop, and dump commands to it.
Additionally, initial documentation is introduced to the README, which
explains how to build and use the tool.

Finally, some minor issues are fixed: proper handling of sampling interval
(in cycles), proper handling of trace duration (in seconds), and printing
stack and method summaries in ascending order for user convenience.

brendangregg commented on README.md:
missing PID

brendangregg commented on README.md:
missing PID (sorry, kept clicking the wrong github button).

brendangregg commented on README.md:
Should mention overhead of this approach. What is the overhead of AsyncGetCallTrace?

One of our major microservices has a java call stack that gets to over 3000 frames deep, 1000 frames after inlining. That's the outlier where -XX:+PreserveFramePointer can add as much as 10% overhead at times (most microservices that overhead is < 0.3%, and close to 0.0%).

Even if the overhead of AsyncGetCallTrace was high, it may still be preferable to always-on frame pointer overhead. But it's important to give some guidance so the end user can make informed decisions...

brendangregg commented on README.md:
PS. It's ok to say the overhead is not yet quantified. :)

apangin commented on README.md:
Not really. PERF_COUNT_SW_CPU_CLOCK counter (used by default) does not require root privileges. async-profiler opens per-thread events that are also available for unprivileged users. Before Linux 4.6 it was even possible to get kernel callchains. Since 4.6 only user-mode profiling is available by default, but it is possible to override this by setting /proc/sys/kernel/perf_event_paranoid to 1 or below. Also /proc/sys/kernel/kptr_restrict should be 0 to allow resolving kernel symbols by non-root.
In fact, I could successfully get full-stack profiles under non-root account on all systems I've tested, e.g. CentOS 7 / 3.10 kernel or Ubuntu 16.04 / 4.4 kernel.

apangin commented on run-command.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sudo java -cp $JAVA_HOME/lib/tools.jar:./build Attach ./build/libasyncProfiler.so "$@"
```
As mentioned above, sudo is not required

goldshtn replied:
Fixed.

apangin commented on src/Attach.java:
```diff
@@ -0,0 +1,50 @@
+            vm.detach();
+        }
+    }
+}
```
I was actually thinking on including [jattach](https://github.com/apangin/jattach) launcher for this purpose, as it has faster startup and does not require tools.jar available in JDK but not JRE.

goldshtn replied:
Oh cool, would you like me to just copy the Linux implementation in and build a standalone tool out of it? Or do you want to tell people they need to download jattach separately (I think it's not as nice in terms of friction)?

apangin replied:
The tool is rather small - it's OK to copy it here.

goldshtn replied:
Sounds good - I will copy it from there.

goldshtn commented on README.md:
Great! Will fix the instructions.

goldshtn commented on README.md:
Makes sense. Will measure and report back.

goldshtn commented on README.md:
Fixed.

goldshtn commented on README.md:
Fixed.

goldshtn commented:
@apangin @incubos @brendangregg Added jattach, reworked the helper script (profiler.sh) to be a little more useful, and rebased to clean up the commits. I think this is good to go, and I will file an issue for the overhead measurement work, which I think should be done separately.

goldshtn commented:
Also, I believe this resolves #1.

apangin commented on README.md:
```diff
@@ -168,3 +168,11 @@ of 127 frames. On recent Linux kernels, this can be configured using
+but you will not see the non-Java frames _preceding_ the Java frames on the
+stack. For example, if `start_thread` called `JavaMain` and then your Java
+code started running, you will not see the first two frames in the resulting
+stack if `perf_events` is unable to unwind the whole stack.
```
This is general limitation of the approach, even with PreserveFramePointer enabled. The native stack ends at the first Java frame; everything below it is obtained via AsyncGetCallTrace.

goldshtn replied:
Fixing, thanks.

apangin commented on src/profiler.cpp:
```diff
@@ -149,8 +149,7 @@ void Profiler::checkDeadline() {
 
-        _running = false;
-        setTimer(0, 0);
+        stop();
```
It is not safe to call Profiler::stop() from within signal handler.
Automatic stop mechanism should be either re-implemented  (e.g. by sleeping in a separate thread) or removed altogether. It's OK to remove it for now and fix later in a separate issue.

goldshtn replied:
Right. I'll get rid of the whole deadline thing for now, then, and create a separate issue to track it, as you suggested.

goldshtn commented:
@apangin Addressed the last two comments, opened an issue about the automatic stop in #15.

apangin commented:
Great! Thanks for all your efforts.
